AlphaWallet/KeyManagement/EtherKeystore.swift

```diff
@@ -0,0 +1,647 @@
+            PromptBackupCoordinator(keystore: self, wallet: wallet, config: .init()).markWalletAsImported()
+        case .failure:
+            break
+        }
+        completion(results)
+    }
+
+    func importWallet(type: ImportType) -> Result<Wallet, KeystoreError> {
+        switch type {
+        case .keystore(let json, let password):
+            guard let keystore = try? LegacyFileBasedKeystore() else {
+                return .failure(.failedToExportPrivateKey)
+            }
+            let result = keystore.getPrivateKeyFromKeystoreFile(json: json, password: password)
+            switch result {
+            case .success(let privateKey):
+                return importWallet(type: .privateKey(privateKey: privateKey))
+            case .failure(let error):
+                return .failure(error)
+            }
+        case .privateKey(let privateKey):
+            let address = AlphaWallet.Address(fromPrivateKey: privateKey)
+            let hasEthereumAddressAlready = wallets.map({ $0.address }).contains {
+                $0.sameContract(as: address)
+            }
+            guard !hasEthereumAddressAlready else {
+                return .failure(.duplicateAccount)
+            }
+            if isUserPresenceCheckPossible {
+                let isSuccessful = savePrivateKeyForNonHdWallet(privateKey, forAccount: .init(address: address), withUserPresence: false)
+                guard isSuccessful else { return .failure(.failedToCreateWallet) }
+                let _ = savePrivateKeyForNonHdWallet(privateKey, forAccount: .init(address: address), withUserPresence: true)
+            } else {
+                let isSuccessful = savePrivateKeyForNonHdWallet(privateKey, forAccount: .init(address: address), withUserPresence: false)
+                guard isSuccessful else { return .failure(.failedToCreateWallet) }
+            }
+            addToListOfEthereumAddressesWithPrivateKeys(address)
+            return .success(Wallet(type: .real(.init(address: address))))
+        case .mnemonic(let mnemonic, _):
+            let mnemonicString = mnemonic.joined(separator: " ")
+            let wallet = HDWallet(mnemonic: mnemonicString, passphrase: emptyPassphrase)
+            let privateKey = derivePrivateKeyOfAccount0(fromHdWallet: wallet)
+            let address = AlphaWallet.Address(fromPrivateKey: privateKey)
+            let hasEthereumAddressAlready = wallets.map({ $0.address }).contains {
+                $0.sameContract(as: address)
+            }
+            guard !hasEthereumAddressAlready else {
+                return .failure(.duplicateAccount)
+            }
+            let seed = HDWallet.computeSeedWithChecksum(fromSeedPhrase: mnemonicString)
+            if isUserPresenceCheckPossible {
+                let isSuccessful = saveSeedForHdWallet(seed, forAccount: .init(address: address), withUserPresence: false)
```

Comment:
Could the flow be simplified? In particular, instead of storing without user presence first, and then trying to store with user presence, why not check for user presence first and then store accordingly?

Reply:
We do check if user presence is available first (but not ask user to verify user presence, otherwise creating each wallet as well as silently migrating old keystore wallets before this PR will require user to verify user-presence).

And if user presence is available (i.e. at least the iOS passcode is set), we save the set that doesn't require user presence first. We don't expect it to fail (but to be sure, we check the result `isSuccessful` the first time to be defensive), and if it works, then we proceed to save the version that requires user-presence. It doesn't matter if the set that requires user presence fails at this point.

This looks a bit tricky but it is because we don't get notified via a hook when user-presence is enabled/disabled. i.e the code handles this case:

1. User deletes iOS passcode
2. Creates a wallet (only k' can be written since there's no passcode)
3. Use the wallet, will attempt to run like (5), but ends up accessing k', as iOS wouldn't/can't prompt to verify user presence because there's no passcode
4. User enables iOS passcode
5. Use the wallet. We know user presence check is available. Tries to access k, fails because it's absent, so we use k' to write k and read k again. This time k works and user is prompted for user presence (as k is read).

For (5), We can't use the result from k' for wallet access right away because we want the user to be prompted to verify user presence, hence we read k again (after writing it, but we are writing k because we don't have it)

Reply:
Thanks for clarifying Boon. The only other comment I have is that it might make sense to change the new comments you added to the code to not talk about a PR, since that is a bit out of context for a code comment.

Reply:
Thanks for clarifying, Boon. The only other comment I have is that it might make sense to change the new comments you added to the code to not talk about a PR, since that is a bit out of context for a code comment.